### ISSUE TITLE
이것 참고해봐... CNN 관련한 부분도 있는 것 같아.

### ISSUE THREAD

**Issue by BluePilgrim** [Skkeem/mnist#2]
SVM으로 하려다보니, 결국 Lagrangian dual problem의 해를 찾는 부분 구현이 핵심인 것 같아서,
이것저것 뒤져보다가 발견한 것...
CNN과 MLP 관련 부분이 포함된 것 같다. Theano 기반으로...

https://groups.google.com/forum/#!topic/theano-users/on4D16jqRX8

**Comment by Skkeem** [comment]
넵

**Comment by BluePilgrim** [comment]
지금 구현할 때 Theano, Tensor를 많이 쓰고 있니?
여기저기 참고해서 SVM 계산하는 부분은 된 것 같은데, input data와 붙이려니, label이나 이런게 tensorVariable 이런 식으로 되어 있어서...

**Comment by Skkeem** [comment]
네 기본적으로 numpy 자료들도 theano.tensor로 변환해서 사용해요. 그거 없이 하셨다면 안쓰셔도 상관없을것 같아요. input_data 쓰시면 numpy 자료형으로 나옵니다

**Comment by BluePilgrim** [comment]
오케이~~~

**Comment by BluePilgrim** [comment]
간단히 SVM 기반의 구현물은 집어넣었다...
아무 생각없이, 28*28을 feature로 다 잡고, 그냥 훈련시켰는데, 아직 55000개를 training 해보지는 않았음.
시간 장난 아니게 걸릴 것 같은데...

프로젝트 설명서에 있는 추가 feature를 구현해야할까 고민 중이다.

**Comment by Skkeem** [comment]
중간에 히든레이어 하나 추가했는데 엄청오래걸리네요.
에러는 1.7프로 정도로 확줄긴합니다.
feature는 프로젝트 1번 말씀하시는건가요? 그거는 svm아니어도 따로 해도 될것같아요.

**Comment by BluePilgrim** [comment]
대단혀~~~ 난 정확도가 얼마일지 잘 모르겠다. 일단 돌리는 것 시작했으니 한참 있다가 결과 나올 듯...

보다 현실적인 것은, 프로젝트 1번 내용처럼 몇가지 대표 feature를 만들거나 PCA를 통해서 좀 복잡도를 낮춘 다음에 이런저런 알고리즘을 돌리는 게 맞을 것 같기는 하다:) 1분정도 PCA를 해볼까 생각하다가 자제하기로 했음^^

**Comment by Skkeem** [comment]
저는 일단 자야할것 같습니다..

**Comment by BluePilgrim** [comment]
몇가지 고쳐서 돌리고 있는데, 언제 끝날지는 모르겠구나.. 발표 전에는 끝나야할텐데...(너무 비효율적으로 계산하는 것 같은 느낌이 들어서^^)

나도 이제 슬슬 졸린다.

1과 7의 구분을 위한 feature set에 대한 것은 여러 가지가 있을 것 같기는 한데, 생각나는 대로 적어보면,
=. center를 중심으로 위아래를 뒤짚었을 경우, 동일한 y 값을 가진 픽셀 사이의 거리. 만약 없을 경우에는 이미지 크기에 해당하는 28을 가정, 이렇게 하여, 거리의 합을 더한다.
=. 두 개 사이의 구분이라, 그냥 이미지 상에서 색깔이 있는(0이 아닌) 픽셀들의 개수
=. 이미지 상에서 가장 왼쪽에 있는 색깔이 있는 픽셀의 x 좌표
=. 수평으로 나눴을 때, 위쪽과 아래쪽의 색깔이 있는 픽셀들의 상대적 비율

뭐 이런 것들이 있을 것 같아...

이것을 plot하려면, 입력으로 주어진 MNIST 데이터를 사용해야 하는걸까?

**Comment by Skkeem** [comment]
네 mnist에서 label 1, 7인거 뽑아 사용해야할것같아요

**Comment by BluePilgrim** [comment]
그냥 Feature는 간단한 걸로 하자.
1) 28x28 상에서 색깔이 있는(non-zero) 픽셀의 개수
2) 28x28 상에서 가장 왼쪽에 있는 픽셀의 x 좌표

그럼 플롯하기가 쉬울 것 같기는 하다.^^

**Comment by Skkeem** [comment]
플롯은 feature마다 나오는 값만 있으면 되니까 어렵지는 않을 것 같아요.

아래 사이트에 나오는 것 같이 하면 좋을 것 같습니다.

http://colah.github.io/posts/2014-10-Visualizing-MNIST/

2015년 12월 16일 오전 11:05, David Yang <notifications@github.com>님이 작성:

**Comment by BluePilgrim** [comment]
ㅋㅋ 뭐가 많구나...^^ 대략 질문에 대한 답이라는 측면에서는 위의 것도 큰 문제는 없을 것 같아.
feature 1 : F1(image) = Sum_j Sum_i [f(c(i, j)], f(c(i,j)) = 1 if c(i,j) > 0 else 0
feature 2 : F2(image) = min_j [argmin_i (c(i,j) > 0)]

두개 feature가 돌려주는 값의 스케일이 달라서 normalize 이슈가 있을 것 같기는 하지만, 문제 해결 용도로는 큰 무리 없을 것 같아.

**Comment by Skkeem** [comment]
http://matplotlib.org/examples/shapes_and_collections/scatter_demo.html
matplotlib 2-d scatter plot 그리는 법

형님 제가 plot 하는 거랑 feature 하나 뽑는 것 해볼게요

혹시 input_data에서 0과 7만 걸러내는 것 만들어주실수 있으세요? label로 거르면 될것같아요

**Comment by BluePilgrim** [comment]
okay.... 0인 것 array와 1인 것 array 두개를 tuple로 넘겨주는 걸로 하면 되겠지?

**Comment by BluePilgrim** [comment]
import input_data





def load_data():
    mnist = input_data.read_data_sets(".")

    print 'Reading data set complete'

    return mnist

def extract_0_and_7(images, labels):
    array0 = []
    array7 = []

    for	i in range(len(labels)):
	if labels[i] == 0:
            array0.append( images[i] )
	elif labels[i] == 7:
            array7.append( images[i] )

    return (array0, array7)

if __name__ == "__main__":


    datasets = load_data()

    #print datasets.validation.images.shape                                                                                                                                                                                                                                                                                   
    #print datasets.validation.labels.shape                                                                                                                                                                                                                                                                                   
    #print datasets.validation.labels                                                                                                                                                                                                                                                                                         

    """                                                                                                                                                                                                                                                                                                                       
    SIZE = 1000                                                                                                                                                                                                                                                                                                               
    train_images = datasets.train.images[:SIZE]                                                                                                                                                                                                                                                                               
    train_labels = datasets.train.labels[:SIZE]                                                                                                                                                                                                                                                                               
                                                                                                                                                                                                                                                                                                                              
    test_images = datasets.test.images[:1000]                                                                                                                                                                                                                                                                                 
    test_labels = datasets.test.labels[:1000]                                                                                                                                                                                                                                                                                 
    """

    train_images = datasets.train.images
    train_labels = datasets.train.labels

    test_images = datasets.test.images
    test_labels = datasets.test.labels

    array0, array7 = extract_0_and_7(train_images, train_labels)

    print "number of 0 = %d, number of 7 = %d" % ( len(array0),	len(array7) )

**Comment by BluePilgrim** [comment]
소스로 올릴께...
뭐 간단한 거니..
